util/reggen/gen_rust.py

```diff
@@ -0,0 +1,388 @@
+    r"""Produces a pub const string. Result includes newline.
+
+    Arguments:
+    name - Name of the constant
+    val - Value of the constant
+    existing_defines - set of already generated define names.
+        Error if `name` is in `existing_defines`.
+
+    Example result:
+    name = 'A_NAME'
+    val = '10'
+
+    pub const A_NAME: u32 = 10
+
+    """
+
+    name = name + suffix
+    if name in existing_defines:
+        log.error("Duplicate pub const for " + name)
+        sys.exit(1)
+
+    define_declare = 'pub const ' + name + ': ' + data_type(name, val)
+
+    oneline_define = define_declare + ' = ' + val + ';'
+
+    existing_defines.add(name)
+
+    return oneline_define + '\n'
+
+
+def gen_const_register(outstr: TextIO,
+                       reg: Register,
+                       comp: str,
+                       width: int,
+                       rnames: Set[str],
+                       existing_defines: Set[str]) -> None:
+    rname = reg.name
+    offset = reg.offset
+
+    genout(outstr, format_comment(first_line(reg.desc)))
+    defname = as_define(comp + '_' + rname)
+    statement = gen_const(defname, '_REG_OFFSET', hex(offset), existing_defines)
+    genout(outstr, statement)
+
+    for field in reg.fields:
+        dname = defname + '_' + as_define(field.name)
+        field_width = field.bits.width()
+
+        if field_width == 1:
+            # single bit
+            statement = gen_const(dname, '_BIT', str(field.bits.lsb), existing_defines)
+            genout(outstr, statement)
+        else:
+            # multiple bits (unless it is the whole register)
+            if field_width != width:
+                mask = field.bits.bitmask() >> field.bits.lsb
+                statement = gen_const(dname, '_MASK', hex(mask), existing_defines)
+                genout(outstr, statement)
+
+                statement = gen_const(dname, '_OFFSET', str(field.bits.lsb), existing_defines)
+                genout(outstr, statement)
+            if field.enum is not None:
+                for enum in field.enum:
+                    ename = as_define(enum.name)
+                    value = hex(enum.value)
+                    statement = gen_const(
+                        defname + '_' + as_define(field.name),
+                        '_VALUE_' + ename,
+                        value,
+                        existing_defines)
+                    genout(outstr, statement)
+    genout(outstr, '\n')
+    return
+
+
+def gen_const_window(outstr: TextIO,
+                     win: Window,
+                     comp: str,
+                     regwidth: int,
+                     rnames: Set[str],
+                     existing_defines: Set[str]) -> None:
+    offset = win.offset
+
+    genout(outstr, format_comment('Memory area: ' + first_line(win.desc)))
+    defname = as_define(comp + '_' + win.name)
+    genout(
+        outstr,
+        gen_const(defname, '_REG_OFFSET', hex(offset), existing_defines))
```

Comment:
I *think* that every callsite of `gen_const` is of the form `genout(outstr, gen_const(...))`. Maybe we should pass `outstr` to `gen_const` and include the `genout` call there? That would remove lots of duplicated code (and awkward formatting) from the call sites.

Reply:
Done.